02-graph_upstream.py

```diff
@@ -75,7 +81,78 @@ def get_version(self, url):
         return str(ver).replace('-', '_')
 
 
-sources = [PyPI(), CRAN(), Github()]
+class NullUndefined(jinja2.Undefined):
+    def __unicode__(self):
+        return self._undefined_name
+
+    def __getattr__(self, name):
+        return '{}.{}'.format(self, name)
+
+    def __getitem__(self, name):
+        return '{}["{}"]'.format(self, name)
+
+
+def next_version(ver):
+    ver_split = []
+    ver_dot_split = ver.split('.')
+    for s in ver_dot_split:
+        ver_dash_split = s.split('_')
+        for j in ver_dash_split:
+            ver_split.append(j)
+            ver_split.append('_')
+        ver_split[-1] = '.'
+    del ver_split[-1]
+    for j in reversed(range(len(ver_split))):
+        try:
+            t = int(ver_split[j])
+        except:
+            continue
+        else:
+            ver_split[j] = str(t + 1)
+            yield ''.join(ver_split)
+            ver_split[j] = '0'
+
+
+class RawURL:
+    name = 'RawURL'
+    def get_url(self, meta_yaml):
+        pkg = meta_yaml['name']
+        url_template = "https://raw.githubusercontent.com/conda-forge/{}-feedstock/master/recipe/meta.yaml"
```

Comment:
This will fail when the feedstock name and the package name don't match. The name in the node attributes matches the package name, the node itself matches the feedstock name. See: https://github.com/regro/cf-scripts/pull/82/files#r173626802